src/channel/mpmc.rs

```diff
@@ -363,51 +379,120 @@ where A: RingBuf<Item=T>
-where A: RingBuf<Item=T>
+impl<MutexType: RawMutex, T, A> ChannelSendAccess<T> for GenericChannel<MutexType, T, A>
+where
+    A: RingBuf<Item = T>,
 {
     unsafe fn try_send(
         &self,
         wait_node: &mut ListNode<SendWaitQueueEntry<T>>,
         cx: &mut Context<'_>,
-    ) -> (Poll<()>, Option<T>)
-    {
+    ) -> (Poll<()>, Option<T>) {
         self.inner.lock().try_send(wait_node, cx)
     }
 
-    fn remove_send_waiter(&self, wait_node: &mut ListNode<SendWaitQueueEntry<T>>)
-    {
+    fn remove_send_waiter(&self, wait_node: &mut ListNode<SendWaitQueueEntry<T>>) {
         self.inner.lock().remove_send_waiter(wait_node)
     }
 }
 
-impl <MutexType: RawMutex, T, A> ChannelReceiveAccess<T> for GenericChannel<MutexType, T, A>
-where A: RingBuf<Item=T>
+impl<MutexType: RawMutex, T, A> ChannelReceiveAccess<T> for GenericChannel<MutexType, T, A>
+where
+    A: RingBuf<Item = T>,
 {
     unsafe fn try_receive(
         &self,
         wait_node: &mut ListNode<RecvWaitQueueEntry>,
         cx: &mut Context<'_>,
-    ) -> Poll<Option<T>>
-    {
+    ) -> Poll<Option<T>> {
         self.inner.lock().try_receive(wait_node, cx)
     }
 
-    fn remove_receive_waiter(&self, wait_node: &mut ListNode<RecvWaitQueueEntry>)
-    {
+    fn remove_receive_waiter(&self, wait_node: &mut ListNode<RecvWaitQueueEntry>) {
         self.inner.lock().remove_receive_waiter(wait_node)
     }
 }
 
+/// A stream that receives from a `GenericChannel`.
+#[derive(Debug)]
+pub struct ChannelStream<'a, MutexType: RawMutex, T, A>
+where
+    A: RingBuf<Item = T>,
+{
+    channel: Option<&'a GenericChannel<MutexType, T, A>>,
+    future: Option<ChannelReceiveFuture<'a, MutexType, T>>,
+}
+
+impl<'a, MutexType, T, A> Stream for ChannelStream<'a, MutexType, T, A>
+where
+    A: RingBuf<Item = T>,
+    MutexType: RawMutex,
+{
+    type Item = T;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        // It might be possible to use Pin::map_unchecked here instead of the two unsafe APIs.
+        // However this didn't seem to work for some borrow checker reasons
+
+        // Safety: The next operations are safe, because Pin promises us that
+        // the address of the wait queue entry inside ChannelReceiveFuture is stable,
+        // and we don't move any fields inside the future until it gets dropped.
+        let mut_self: &mut Self = unsafe { Pin::get_unchecked_mut(self) };
+        let channel = mut_self.channel.unwrap();
+        if mut_self.future.is_none() {
+            mut_self.future.replace(channel.receive());
+        }
+        let fut = mut_self.future.as_mut().unwrap();
+
+        // Safety: Here we guarantee that the pinned future will not move until
+        // it resolves.
+        let poll = unsafe {
+            let pin_fut = Pin::new_unchecked(fut);
+            pin_fut.poll(cx)
+        };
+
+        // Future has not yet resolved, keep it.
```

Comment:
Maybe add: "Otherwise drop it"

Otherwise the comment is exactly the opposite of what happens below

Reply:
Typo : `if Future has not yet resolved, keep it.`